text/0000-pi-types.md

```diff
@@ -0,0 +1,1139 @@
+- Feature Name: pi-types
+- Start Date: 2016-06-22
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+We propose a simple, yet sufficiently expressive, addition of dependent-types
+(also known as, Π-types and value-types).
+
+Type checking will not require SMT-solvers or other forms of theorem provers.
+
+## Generic value parameters
+
+A `const` type parameter acts like a generic parameter, containing a constant
+expression. Declaring a generic parameter `const a: usize`, declares a constant
+variable `a` of type `usize`.
+
+One can create implementations, structs, enums, and traits, abstracting over
+this generic value parameter.
+
+Such a parameter acts type-like in the context of types, generics, and
+polymorphism, and value-like in the context of expressions, function bodies,
+and applications.
+
+## Compile time calculations on constant parameters
+
+Since it is simply consisting of constexprs, one can apply constant functions
+(`const fn`) to the parameter, to perform compile time, type level calculations
+on the parameter. This allows for great expressiveness as `const fn` improves.
+
+## Expression `where` bounds
+
+The second construct added is the constant expression in `where` bounds. These
+contains statements about the constant parameters, which are checked at compile
+time.
+
+## Type checking
+
+Type checking is done by using a [transitive rule](#transitivity-of-bounds), such that `where` bounds must
+be implied by the caller.
+
+# Motivation
+[motivation]: #motivation
+
+An often requested feature is the "type-level numerals", which enables generic
+length arrays. The current proposals are often limited to integers or even lack
+of value maps, and other critical features.
+
+There is a whole lot of other usecases as well. These allows certain often
+requested features to live in standalone libraries (e.g., [bounded-integers](#bounded-integersinterval-arithmetics),
+[type level numerals](#array-generics), [statically checked indexing](#statically-checked-indexing), lattice types).
+
+It allows for creating powerful abstractions without type-level hackery.
+
+## What we want, and what we don't want
+
+We have to be very careful to avoid certain things, while still preserving the core features:
+
+1. Ability to use, manipulate, and constraint values at type-level.
+2. The ability to use said values on expression-level (runtime).
+
+Yet, we do not want:
+
+1. SMT-solvers, due to not only undecidability (note, although, that SAT is
+   decidable) and performance, but the complications it adds to `rustc`.
+2. Monomorphisation-time errors, i.e. errors that happens during codegen of
+   generic functions. We try to avoid adding _more_ of these (as noted by
+   petrochenkov, these [already exists](https://github.com/rust-lang/rfcs/pull/1657#discussion_r68202733))
```

Comment:
@torpak: Consider `mmap(2)` (which allows any file to be addressed as memory), and that there are indeed array-like datasets that large.